### PR TITLE
Add `finallist` command too bootstrap process

### PR DESCRIPTION
**Pull request by danuio**
Resolves https://github.com/dapperlabs/flow-go/issues/4970.

This PR adds a new command `finallist` to the bootstrap command line interface. The command generates the final list of nodes to be used by the `finalize` command. 

```console
./cmd/bootstrap/cmd/finallist.go \ 
    --partner-infos ./mainnet1 \
    --flow-infos ./flow_infos \
    --staking-nodes ./staking_nodes \ 
    -o ./node-config.json
```

Also changes the `readPartnerNodes` method in `finalize` command to return a `NodePubInfo` rather than ` PartnerNodePubInfo` as the partner nodes read in are the output of the `pull` command which has stake 0 for all the nodes.

### CODE REVIEW

File: cmd/bootstrap/cmd/final_list.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+		// validate every single partner node
+		nodeID := validateNodeID(staking.NodeID)
+		networkPubKey := validateNetworkPubKey(staking.NetworkPubKey)
+		stakingPubKey := validateStakingPubKey(staking.StakingPubKey)
+
+		node := model.NewPublicNodeInfo(
+			nodeID,
+			staking.Role,
+			staking.Address,
+			0,
```
Comment (vishalchangrani) [inline cmd/bootstrap/cmd/final_list.go:82]: wouldn't this be a valid stake? and not 0?

File: cmd/bootstrap/cmd/final_list.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	// check node id mismatch
+	// check node type mismathc
+}
+
+func assembleInternalNodesWithoutStake() []model.NodeInfo {
+	privInternals := readInternalNodes()
+	log.Info().Msgf("read %v internal private node-info files", len(privInternals))
+
+	var nodes []model.NodeInfo
+	for _, internal := range privInternals {
```
Comment (vishalchangrani) [inline cmd/bootstrap/cmd/final_list.go:103]: you could commonize this for loop (its the same in the all three functions)

File: cmd/bootstrap/cmd/final_list.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+var (
+	flagStakingNodesDir string
+)
+
+// finallistCmd represents the final list command
+var finalListCmd = &cobra.Command{
+	Use:   "finallist",
+	Short: "",
+	Long:  ``,
```
Comment (vishalchangrani) [inline cmd/bootstrap/cmd/final_list.go:20]:
```suggestion
	Long:  "",
```

File: cmd/bootstrap/cmd/final_list.go (hunk 1 above)
Comment (danuio) [reply cmd/bootstrap/cmd/final_list.go]: Yes thanks!

File: cmd/bootstrap/cmd/final_list.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+	mixedNodeInfos := mergeNodeInfos(flowNodes, partnerNodes)
+
+	// reconcile nodes from staking contract nodes
+	reconcileNodes(mixedNodeInfos, stakingNodes)
+
+	// write node-config.json with the new list of nodes to be used for the `finalize` command
+	writeJSON(fmt.Sprintf(flagOutdir, "node-config.json"), model.ToPublicNodeInfoList(mixedNodeInfos))
+}
+
+func reconcileNodes(nodes []model.NodeInfo, stakingNodes []model.NodeInfo) {
```
Comment (Kay-Zee) [inline cmd/bootstrap/cmd/final_list.go:69]: This just compares the two lists, and warns if any field mis-matches?

**Review by Kay-Zee — APPROVED** [review]

File: cmd/bootstrap/cmd/final_list.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+	// check node count
+	if len(nodes) != len(stakingNodes) {
+		log.Error().Int("nodes", len(nodes)).Int("staked nodes", len(stakingNodes)).
+			Msg("staked node count does not match flow and parnter node count")
+	}
+
+	var nodesByID map[flow.Identifier]model.NodeInfo
+	for _, node := range nodes {
+		nodesByID[node.NodeID] = node
+	}
```
Comment (zhangchiqing) [inline cmd/bootstrap/cmd/final_list.go:79]: @danuio let's make the check a bit more strict.

1. we need to go through each list, and check if there is missing node id on _each side_. If there is, print all the missing node id and then exist the program.

When we found missing node id during sporking, we could fix the data, and run this cmd again, until it passes.

2. We need to exist if the role doesn't match and print error

3. We could print warning if address or networking/staking key doesn't match, but still the program continue, as we mentioned, we will only strictly match node id for now.

**Review by zhangchiqing — CHANGES_REQUESTED** [review]

File: cmd/bootstrap/cmd/final_list.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+func init() {
+	rootCmd.AddCommand(finalListCmd)
+	addFinalListFlags()
+}
+
+func addFinalListFlags() {
+	// partner node info flag
+	finalListCmd.Flags().StringVar(&flagPartnerNodeInfoDir, "partner-infoss", "", "path to a directory containing all parnter nodes details")
```
Comment (zhangchiqing) [inline cmd/bootstrap/cmd/final_list.go:32]:
```suggestion
	finalListCmd.Flags().StringVar(&flagPartnerNodeInfoDir, "partner-infos", "", "path to a directory containing all parnter nodes details")
```

File: cmd/bootstrap/cmd/final_list.go [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+// validateNodeIDs will go through both sets of nodes and ensure that no node-id
+// are missing. It will log all missing node ID's and throw an error.
+func validateNodeIDs(collectedNodes []model.NodeInfo, stakedNodes []model.NodeInfo) {
+
+	// go through staking nodes
+	invalidStakingNodes := make([]model.NodeInfo, 0)
+	for _, node := range stakedNodes {
+		if node.NodeID.String() == "" {
```
Comment (zhangchiqing) [inline cmd/bootstrap/cmd/final_list.go:147]: Sorry, maybe I was clear. The validation is not only to check whether NodeID is missing, but also see if there is collected node which is not a staked node, and if there is a staked node which is not a staked nodes. 

For instance say collectedNodes has nodes with ID `A, B, C`, and stakedNodes has `B, C, D`, and then we have mismatching collected nodes `A` and mismatching staked nodes `D`.

If there is either mismatching collected nodes or mismatching staked nodes, the program will print fatal error with the miss matching nodes.

Does it make sense?

Comment (danuio) [reply cmd/bootstrap/cmd/final_list.go:147]: Ahh yes I see. Ok my bad,  I'll fix that now.

**Review by zhangchiqing — APPROVED** [review]

**Pull request merged**